### DISCUSSION TITLE
[sparse matrix] how to jaxify the sparse matrix

### DISCUSSION THREAD

**Discussion by JiahaoYao** [jax-ml/jax#10422] [Q&A] [answered]
```python
import scipy
import scipy.sparse
from scipy.sparse import random as sp_rand
from tqdm import tqdm, trange
import jax 
from jax.experimental import sparse

my_sp_mat_list = [ sp_rand(2000, 2000, density=0.01) for i in trange(500)] 

my_sp_mat_list_jax = [sparse.BCOO.fromdense( H.toarray() ) for H in tqdm(my_sp_mat_list)]
```

I wonder if there is a way to directly convert the sparse matrix in `numpy` to `jax.DeviceArray`? Because the current way of first converting the numpy sparse matrix to dense and then convert to the sparse matrix seems costly.

**Comment by jakevdp** [comment, marked as answer]
You can use `BCOO.from_scipy_sparse`:
```python
my_sp_mat_list_jax = [sparse.BCOO.from_scipy_sparse(H) for H in my_sp_mat_list]
```

**Reply by JiahaoYao** [reply]
Hi @jakevdp , Awesome! Thanks for your prompt reply. 

Here is a follow-up question: 

```python
import scipy
import scipy.sparse
from scipy.sparse import random as sp_rand
from tqdm import tqdm, trange
import jax 
from jax.experimental import sparse
import jax.random as jrd 
import jax.numpy as jnp 
from jax import vmap 

my_sp_mat_list = [ sp_rand(2000, 2000, density=0.01) for i in trange(20)] 

my_sp_mat_list_jax = [sparse.BCOO.from_scipy_sparse( H ) for H in tqdm(my_sp_mat_list)]

key = jrd.PRNGKey(42)
vec = jrd.normal(key, (2000,))

def matvec(mat, vec): 
    return mat @ vec 

f_sp = sparse.sparsify(matvec)

vec_list = [ f_sp(H, vec) for H in tqdm(my_sp_mat_list_jax)]
```

However, a `jax-style` implementation should use `vmap` like this: 
```python
f_sp_vmap = vmap(matvec, (0, None))
vec_list = f_sp_vmap(my_sp_mat_list_jax, vec)
```
However, this does not work because vmap does not work for the list. In fact, if there is something like `jnp.concatenate` here, I can `my_sp_mat_list_jax = jnp.concatenate(my_sp_mat_list_jax, axis=0)` so that I can use `axis=0` as the batch axis for the `vmap`. 

Or do you know some alternative way to implement this? Thank you so much again!

**Reply by JiahaoYao** [reply]
Just to soft bump this up, thanks!

**Reply by jakevdp** [reply]
Sorry for the delay – I take my weekends seriously 😀 .

Just as vmap will not work on lists of device arrays, vmap will not work on lists of sparse arrays. What you'll need is to concatenate those. Unfortunately we've not yet implemented the sparsification rule for `lax.concatenate_p`, but it should be doable. Let me see if I can cook up a short example.

**Reply by jakevdp** [reply]
Here's a very simple sparse matrix concatenation routine that works for your example. I tried to use asserts to make clear which cases would require more logic to implement correctly:
```python
from jax import lax

def concat(arr_list, axis=0):
  assert len(arr_list) > 0, "Cannot concatenate zero arrays."
  assert axis == 0, "only axis=0 is implemented"
  assert len(set(tuple(a.shape) for a in arr_list)) == 1, "all arrays must have the same shape"
  assert len(set(a.nse for a in arr_list)) == 1, "All entries must have the same nse"
  assert all(a.n_sparse == a.ndim for a in arr_list), "Only sparse dimensions supported"

  data = lax.concatenate([a.data[None] for a in arr_list], dimension=0)
  indices = lax.concatenate([a.indices[None] for a in arr_list], dimension=0)

  return sparse.BCOO((data, indices), shape=(len(arr_list), *arr_list[0].shape))

stacked = concat(my_sp_mat_list_jax)

result = vmap(matvec, in_axes=(0, None))(stacked, vec)
```
Perhaps we can add a sparse rule for `lax.concatenate` that does something along these lines.

**Reply by JiahaoYao** [reply]
This is AWESOME, thanks @jakevdp for the prompt reply!

**Reply by jakevdp** [reply]
#10466 adds a sparsification rule for concatenation, and #10454 makes added dimensions default to batch. Once these two changes land, you should be able to do something like this:
```python
my_sp_mat_list_jax = [sparse.BCOO.from_scipy_sparse( H ) for H in tqdm(my_sp_mat_list)]
stacked = sparse.sparsify(jnp.stack)(my_sp_mat_list_jax)
```

**Reply by JiahaoYao** [reply]
Awesome! as always, thanks for being so supportive! Thanks!
